### ISSUE TITLE
Wish: DS shortcut for creating dictionaries with syntax {keys:vals}

### ISSUE THREAD

**Issue by josols** [DynamoDS/Dynamo#8759]
If this issue is with Dynamo for Revit, please post your issue on the [Dynamo for Revit Issues page](https://github.com/DynamoDS/DynamoRevit/issues).

If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue.

## Dynamo version

2.0.0 stable

## Operating system

Win 7

## What did you do? 

Typed {keys,vals} in a codeblock and expected to be able to define a dictionary with variables instead of typing DesignScript.Builtin.Dictionary.ByKeysValues(keys,vals);

## What did you expect to see?

a dictionary.

## What did you see instead?

(Fill in here)
![image](https://user-images.githubusercontent.com/7080648/38869549-5b2e4596-424b-11e8-8aba-abbd81afc5d2.png)

But the values of a dictionary can be made a variable:
![image](https://user-images.githubusercontent.com/7080648/38869570-7734664e-424b-11e8-8635-4c1a30821131.png)

**Comment by jnealb** [comment]
@aparajit-pratap Can you please comment? @Racel fyi.

**Comment by Amoursol** [comment]
@ksobon , @pboyer gave a detail response to my query of non-string key's located here:

[(https://github.com/DynamoDS/Dynamo/issues/8497#issuecomment-361332513)]

**Comment by ksobon** [comment]
So I am guessing the answer is we are not supporting any other Type for dictionary keys because it can confuse people. I might not entirely agree with that, but let it be. @kronz do we close this?

**Comment by mjkkirschner** [comment]
@ksobon also I believe the thought process was that complicates the implementation, and lowers its maintainability for a feature we are not sure a large number of users will take advantage of.

**Comment by josols** [comment]
For me this was not about using strings as dictionary keys. I'm fine about that. But, for me, and others as well, the shortcut in Dynamo 1.x for dictionaries, ie dict[keys]=vals was powerful and easy to use. Could it not be possible to have variables as keys, but only allow the variable to be a string? 

{keys:vals} would be so much easier to use than DesignScript.Builtin.Dictionary.ByKeysValues(keys,vals)...

**Comment by ksobon** [comment]
@mjkkirschner completely understood. That's why I am saying I am game with us just saying `string` is the only type supported. 
@josols that's a valid concern. So long as the variable was of `string` type in my opinion you should be able to define a dictionary like you tried. @mjkkirschner do you agree?

**Comment by gregmarr** [comment]
There is rationale in the PR by @pboyer for the current behavior.  https://github.com/DynamoDS/Dynamo/pull/8551

**Comment by gregmarr** [comment]
>simplifying the parser and eliminating the complexities of interaction with other language features (notably replication), this approach makes our initialization syntax essentially identical to JSON.

**Comment by gregmarr** [comment]
That's not to say that we can't revisit this in the future, but that's why it is the way it is right now.

**Comment by aparajit-pratap** [comment]
@josols @ksobon we have experimented with this idea and found that it could lead to confusing results when a syntax like `{keys : vals}` interferes with other language features of DesignScript like replication. The following example was one such experiment (note that this was before the list `[]` syntax changes): 
![image](https://user-images.githubusercontent.com/5710686/38946805-96e27248-4308-11e8-8b3f-ffffaf8a9ce5.png)
This produced a result where: `[ { "key1" : [ 23, 324 ] } ]` which was hard to reason about and it also differed from the UI node result. 

Then there could be other confusing cases like this simple statement: `a = ["foo", "bar"]; b = { a : "baz" }`. What would one expect this result to be.

We finally concluded that it would not be very useful unless we actually went to the extent of supporting arbitrary expressions and having replication syntax etc., not just identifiers. We didn't want to go the route of supporting arbitrary expressions for the sake of simplicity.

We _could_ extend dictionary keys to support arbitrary expressions in the future but we will also have to ensure that the interaction with other language features is consistent and intelligible at the cost of adding complexity (not to mention the time and effort invested implementing this) rather than making the system a little less powerful and simple to understand. Further, there is always an alternate way to address this by using the `Dictionary.ByKeysValues(keyList, valueList)` method, which is not that much of a stretch.
